drivers/adc/adc_ad5592.c

```diff
@@ -0,0 +1,255 @@
+
+#include <zephyr/drivers/adc.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/drivers/mfd/ad5592.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(adc_ad5592, CONFIG_ADC_LOG_LEVEL);
+
+#define AD5592_ADC_RESOLUTION 12U
+#define AD5592_ADC_MAX_VAL 4096
+
+struct adc_ad5592_config {
+	const struct device *mfd_dev;
+};
+
+struct adc_ad5592_data {
+	struct adc_context ctx;
+	const struct device *dev;
+	uint8_t adc_conf;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+	uint8_t channels;
+	struct k_thread thread;
+	struct k_sem sem;
+
+	K_KERNEL_STACK_MEMBER(stack, CONFIG_ADC_AD5592_ACQUISITION_THREAD_STACK_SIZE);
+};
+
+static int adc_ad5592_channel_setup(const struct device *dev,
+				    const struct adc_channel_cfg *channel_cfg)
+{
+	const struct adc_ad5592_config *config = dev->config;
+	struct adc_ad5592_data *data = dev->data;
+
+	if (channel_cfg->channel_id >= AD5592_PIN_MAX) {
+		LOG_ERR("invalid channel id %d", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	data->adc_conf |= BIT(channel_cfg->channel_id);
+
+	return mfd_ad5592_write_reg(config->mfd_dev, AD5592_REG_ADC_CONFIG, data->adc_conf);
+}
+
+static int adc_ad5592_validate_buffer_size(const struct device *dev,
+					   const struct adc_sequence *sequence)
+{
+	uint8_t channels;
+	size_t needed;
+
+	channels = POPCOUNT(sequence->channels);
+	needed = channels * sizeof(uint16_t);
+
+	if (sequence->buffer_size < needed) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_ad5592_start_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	struct adc_ad5592_data *data = dev->data;
+	int ret;
+
+	if (sequence->resolution != AD5592_ADC_RESOLUTION) {
+		LOG_ERR("invalid resolution %d", sequence->resolution);
+		return -EINVAL;
+	}
+
+	if (find_msb_set(sequence->channels) > AD5592_PIN_MAX) {
+		LOG_ERR("invalid channels in mask: 0x%08x", sequence->channels);
+		return -EINVAL;
+	}
+
+	ret = adc_ad5592_validate_buffer_size(dev, sequence);
+	if (ret < 0) {
+		LOG_ERR("insufficient buffer size");
+		return ret;
+	}
+
+	data->buffer = sequence->buffer;
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_ad5592_read_channel(const struct device *dev, uint8_t channel, uint16_t *result)
+{
+	const struct adc_ad5592_config *config = dev->config;
+	uint16_t val;
+	int ret;
+
+	ret = mfd_ad5592_write_reg(config->mfd_dev, AD5592_REG_SEQ_ADC, BIT(channel));
+	if (ret < 0) {
+		return ret;
+	}
+
+	/*
+	 * Invalid data:
+	 * See Figure 46. Single-Channel ADC Conversion Sequence.
+	 * The first conversion result always returns invalid data.
+	 */
+	(void) mfd_ad5592_read_raw(config->mfd_dev, &val);
+
+	ret = mfd_ad5592_read_raw(config->mfd_dev, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val = sys_be16_to_cpu(val);
+	if (channel >= 1) {
+		val -= channel * AD5592_ADC_MAX_VAL;
+	}
```

Comment:
nit: If this is masking out the ADC channel information from the result field, it might be clearer to just select the lower 12 bits here.

e.g:
```suggestion
	val &= BITMASK(AD5592_ADC_RESOLUTION);
```